File: app/src/main/java/com/example/changfeng/taptapword/ClipboardService.java
package com.example.changfeng.taptapword;

import android.app.Service;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.os.IBinder;
import android.widget.Toast;

import java.util.regex.Pattern;

public class ClipboardService extends Service {

    private static final String TAG = "ClipboardService";

    public static final String clipboardText = "clipboardText";

    private ClipboardManager cb;

    public ClipboardService() {
    }


    @Override
    public void onCreate() {
        super.onCreate();
    }

    @Override
    public IBinder onBind(Intent intent) {
        // TODO: Return the communication channel to the service.
        throw new UnsupportedOperationException("Not yet implemented");
    }


    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        cb = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        cb.setPrimaryClip(ClipData.newPlainText("", ""));
        cb.addPrimaryClipChangedListener(cbListener);
        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        cb.removePrimaryClipChangedListener(cbListener);
        MyLog.d(TAG, "onDestroy()");
    }


    ClipboardManager.OnPrimaryClipChangedListener cbListener = new ClipboardManager.OnPrimaryClipChangedListener() {
        @Override
        public void onPrimaryClipChanged() {
            if (cb.hasPrimaryClip()) {
                for (int i = 0; i < cb.getPrimaryClip().getItemCount(); i++) {

                    if (cb.getPrimaryClip().getItemAt(i).getText() != null && cb.getPrimaryClip().getItemAt(i).getText().length() > 0) {
                        if (!cb.getPrimaryClip().getItemAt(i).getText().toString().trim().isEmpty()) {
                            String clipboardItem = cb.getPrimaryClip().getItemAt(i).getText().toString().trim();

                            if (!Utils.isEnglishWord(clipboardItem)) {
                                break;
                            }

                            Intent intent = new Intent(getApplicationContext(), ConsultWordActivity.class);
                            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            intent.putExtra(clipboardText, clipboardItem);
                            intent.putExtra(ConsultWordActivity.TYPE, ConsultWordActivity.TYPE_COPY);
                            startActivity(intent);
                            break;
                        }
                    }
                }
            }
        }
    };

    void showToast(String info) {
        Toast.makeText(getApplicationContext(), info, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/example/changfeng/taptapword/ClipboardService.kt
package com.example.changfeng.taptapword

import android.app.Notification
import android.app.Service
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.IBinder
import com.example.changfeng.taptapword.util.LogUtils
import com.example.changfeng.taptapword.util.isEnglishWord
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.jetbrains.anko.intentFor
import org.jetbrains.anko.newTask
import android.app.NotificationManager
import android.R
import android.annotation.TargetApi
import android.app.NotificationChannel
import android.graphics.Color
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat.getSystemService


class ClipboardService : Service(), ClipboardManager.OnPrimaryClipChangedListener {

    private val cb: ClipboardManager by lazy {
        getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    }

    override fun onCreate() {
        super.onCreate()
        LogUtils.LOGI(TAG, "onCreate()")
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startMyOwnForeground()
        } else {
            startForeground(1, Notification())
        }
    }

    override fun onBind(intent: Intent): IBinder? {
        throw UnsupportedOperationException("Not yet implemented")
    }


    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        cb.addPrimaryClipChangedListener(this)
        return super.onStartCommand(intent, flags, startId)
    }

    override fun onDestroy() {
        LogUtils.LOGI(TAG, "onDestroy()")
        cb.removePrimaryClipChangedListener(this)
        super.onDestroy()
    }


    override fun onPrimaryClipChanged() {
        LogUtils.LOGI(TAG, "onPrimaryClipChanged() ${cb.hasPrimaryClip()}")
        if (!cb.hasPrimaryClip()) {
            return
        }

        for (i in 0 until cb.primaryClip!!.itemCount) {
            LogUtils.LOGI(TAG, "onPrimaryClipChanged: ${cb.primaryClip!!.getItemAt(i).text}")
        }

        for (i in 0 until cb.primaryClip!!.itemCount) {
            val item = cb.primaryClip!!.getItemAt(i)
            val text = item.text.toString()

            if (!text.isNullOrEmpty()) {
                if (text.contains("changfeng")) {
                    continue
                }

                if (!text.isEnglishWord()) {
                    LogUtils.LOGI(TAG, "onPrimaryClipChanged() $text is not english word")
                    continue
                }
                LogUtils.LOGI(TAG, "onPrimaryClipChanged() start activity")
                startActivity(intentFor<ConsultWordActivity>(clipboardText to text, ConsultWordActivity.TYPE to ConsultWordActivity.TYPE_COPY).newTask())
                return
            }
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun startMyOwnForeground() {
        val channelName = "My Background Service"
        val chan = NotificationChannel(notificationId, channelName, NotificationManager.IMPORTANCE_NONE)
        chan.lightColor = Color.BLUE
        chan.lockscreenVisibility = Notification.VISIBILITY_PRIVATE
        val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        manager.createNotificationChannel(chan)

        val notificationBuilder = NotificationCompat.Builder(this, notificationId)
        val notification = notificationBuilder.setOngoing(true)
                .setContentTitle("单词忍者")
                .setContentText("单词忍者正在后台运行")
                .setPriority(NotificationManager.IMPORTANCE_MIN)
                .setCategory(Notification.CATEGORY_SERVICE)
                .build()
        startForeground(2, notification)
    }

    companion object {
        private const val TAG = "ClipboardService"
        var notificationId = "channelId_tap_tap_word"
        const val clipboardText = "clipboardText"
    }
}
